### PR TITLE
JP-3491: Modify MIRI LRS WCS mapping code

### LINKED ISSUE #8133: assign_wcs introduces a tilt to a line of constant wavelength
_Issue [JP-3491](https://jira.stsci.edu/browse/JP-3491) was created on JIRA by [Greg Sloan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gcsloan):_

The current LRS distortion solutions keep a line of constant wavelength parallel to the row axis on the detector, but assign_wcs tilts the wavelengths in detector coordinates, such that the wavelength changes from one column to the next along a given row.   That leads to erroneous offsets in wavelength depending on position in the cross-dispersion (i.e., spatial) direction in the spectral image.  It also triggers additional bugs, including pixels with NaN values and DQ flags set to 0 instead of 1 (use instead of do_not_use), which in turn triggers other bugs.  For one known dependent bug that has now been repaired, see JIRA ticket JP-3473.

### PR DESCRIPTION
**Pull request by drlaw1558**
Resolves [JP-3491](https://jira.stsci.edu/browse/JP-3491) for unintended tilts in the MIRI LRS wavelength solution by changing the approach to LRS distortion fitting.

Closes #8133 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Review by jemorrison — APPROVED** [review]
To test this and understand it - I took the new code and put it in a python notebook along with reading in the specwcs and plotted each step. I need to do that to understand what was going on.
The code works well in using the specwcs reference file to map what they wavelengths of the pixels are

File: jwst/assign_wcs/miri.py [hunk 1]
```diff
@@ -289,57 +289,29 @@ def lrs_distortion(input_model, reference_files):
...
-    wmax = np.amax(wavereference)
-    wgrid = np.arange(wmin, wmax, dw)
-    # Evaluate the rotated y locations of the grid
-    ygrid = spl2(wgrid)
-    # ygrid now contains the rotated y pixel locations corresponding to
-    # regularly-sampled wavelengths, create the model
-    wavemodel.inverse = models.Tabular1D(lookup_table=ygrid, points=wgrid, name='waverefinv',
-                                         bounds_error=False, fill_value=np.nan)
+    # Fit for the wavelength as a function of Y
+    # Reverse the vectors to that yinv is increasing (needed for spline fitting function)
```
Comment (hbushouse) [inline jwst/assign_wcs/miri.py:307]:
```suggestion
    # Reverse the vectors so that yinv is increasing (needed for spline fitting function)
```

**Review by hbushouse — COMMENTED** [review]
Can't really say much regarding whether this is the right thing to do or not, but the code looks OK.

**Comment by hbushouse** [comment]
Regression tests started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1376

File: jwst/assign_wcs/miri.py (hunk 1 above)
Comment (drlaw1558) [reply jwst/assign_wcs/miri.py]: Fixed.

**Comment by hbushouse** [comment]
Regtest run shows failures for two MIRI LRS tests, with small differences in S_REGION, wavelength, and resulting data array values, which I assume are expected due to these changes.

**Review by hbushouse — APPROVED** [review]

**Comment by hbushouse** [comment]
@nden any thoughts on this? For example, are inverse transforms covered properly, etc.

**Comment by drlaw1558** [comment]
One thing worth thinking about is whether we should add an intermediate WCS frame for the slit, which could help future cube building from LRS data.  Thoughts @jemorrison ?

**Comment by jemorrison** [comment]
I think it might be useful to add a slit frame transform. That would help in the future when building slit-stepped LRS data. It might also help now in creating a resampled LRS output frame from dithered data.

**Comment by hbushouse** [comment]
@drlaw1558 Do you want to add slit frame transforms yet in this PR, or do that separately?

**Comment by drlaw1558** [comment]
After looking at it a little bit, I think that will be somewhat tedious to implement the way that would be most useful and thus should probably be done separately so that it doesn't hold up the original distortion change prompting the PR.  It'll require some rejiggering of the functions and would be useful to test independently.

**Pull request merged**